### PR TITLE
Adds hardcore random mode: Earn your way onto a score-board with difficult random characters

### PR DESCRIPTION
**Pull request by ghost**
:cl: Qustinnus & MrDoomBringer
 add: Adds hardcore random mode, you get a random character with some random bad quirks. You can earn hardcore random score and get on the leaderboards if you survive and get on the shuttle. If the odds are stacked against you more, you will also earn more points.
 add: Gold wheelchair, only for people with 5k hardcore random mode points
/:cl:

Adds an opt-in mode which adds some extra challenge to the game by giving you negative quirks, with the bonus that you get to have clout by going on a leaderboard. This mode exist as somewhat of an extra challenge for veteran players who would like to be cool by being on the leaderboard.

Your character gets a random budget of points to pick quirks with, the more points you end up getting the more score you will get. If you are an antag you will get double the points for making it through (Blind nuke ops win maybe?)

example vid: https://streamable.com/rbnum6 (This one is a super high value run)

I think later on we can tie some small awards for players that do manage to get to certain scores on the scoreboard (VERY small things, like a neat badge or something)

### CODE REVIEW

**Comment by wesoda25** [comment]
bro i kinda like when people take negative quirks for an interesting character and not for a high score

**Comment by MrDoomBringer** [comment]
@wesoda25 i dont think that affects the actual rounds that much though, we'll still be seeing epic meme characters and so on. Just because there's a neat bonus doesn't really make it worse

**Comment by Qustinnus** [comment]
what mrdoombringer says. if people enjoy having bad quirks because it fits their characters thats unchanged, this is more of an extreme "hey what if my character rolled only 1's on stat creation" kind of challenge.

**Comment by wesoda25** [comment]
ik im just whining this is still pretty cool

**Comment by Qustinnus** [comment]
okay if you get to 5k points you now unlock a gold wheelchair for when you're crippled. god speed.

File: code/datums/achievements/hardcore_random.dm
Comment (ATH1909) [inline code/datums/achievements/hardcore_random.dm:3]:
```suggestion
	desc = "Well, I might be a blind, deaf, crippled guy, but hey, at least I'm alive."
```

**Comment by zxaber** [comment]
Wiz and nukie antags shouldn't be a part of this (I'm pretty sure quirks already don't get applied to these two roles). Otherwise, seems neat.

The double points as antag should include greentexting and not just escape. There's nothing really interesting about a blind criple tot sitting in departures all round for the double point bonus.

**Comment by Ziiro** [comment]
going to steal a golden wheelchair from these utter champions to feed the science protolathe

**Comment by itseasytosee** [comment]
Lings can heal themselves out of most disabilities, so this shouldn't apply to them.

**Comment by Shadowflame909** [comment]
also make traitor bros not apply to this.

Team antags will get yeeted hard by this lol

**Comment by itseasytosee** [comment]
So, in reality, this should only count for round start traitor

**Comment by Cartographer-D** [comment]
The BUNDY nuclear operative family gets stuck with Roy, the deaf, blind, crippled man with a brain tumor and a fear of guns.

**Comment by Mickyan** [comment]
If you're going to add rewards for this (which is a dangerous slippery slope that I don't care for but whatever) they should only show up when you have the hardcore option enabled

**Comment by Qustinnus** [comment]
if you turn on hardcore random mode and get a traitor thats part of the game and playing with a crippled guy is part of the fun. I'll look into disabling ling and wiz though.

**Comment by itseasytosee** [comment]
Nukie + wheelchair + arm implant

**Comment by Shadowflame909** [comment]
> 
> 
> The BUNDY nuclear operative family gets stuck with Roy, the deaf, blind, crippled man with a brain tumor and a fear of guns.

whats with the downvotes...do people really want to get stuck with the above, and literally have to carry them

**Comment by trollbreeder** [comment]
Time for wheelchair ops

**Comment by trollbreeder** [comment]
image
Wheelchair ops

**Comment by Qustinnus** [comment]
people can play as a braindead grug nukeop now anyways

**Comment by Qustinnus** [comment]
do quirks even apply to wizard etc right now? it doesnt seem like it does

File: code/controllers/subsystem/processing/quirks.dm
Comment (spookydonut) [inline code/controllers/subsystem/processing/quirks.dm:55]: error: undefined field: "hardcore_survival_score" on /mob/living/carbon

File: code/controllers/subsystem/processing/quirks.dm
Comment (spookydonut) [inline code/controllers/subsystem/processing/quirks.dm:54]:
```suggestion
		var/mob/living/carbon/human/human = user
```

**Comment by anconfuzedrock** [comment]
the warden spawning blind isn't just a downside for the warden it's a downside for the whole fucking security team we should never incentivise this

**Comment by Shadowflame909** [comment]
When you roll traitor bros and you gotta hijack, but your partner is blind, paraplegic, and hallucinating.

Adding onto other peoples suffering isnt fun

**Comment by Qustinnus** [comment]
okay shadowflame for the tenth time; this PR does not change the availability of quirks for roles like nuke ops and wizard (they dont get quirks). for blood bros, nothing changes either. If they had quirks available before they still do now. If you dont want quirks available for blood bros; make a PR for that. this change is completely unrelated to quirk availability to antags.

**Comment by Qustinnus** [comment]
if someone wants to spawn as a blind warden they can do so with or without this PR.

File: code/__HELPERS/roundend.dm
Comment (spookydonut) [inline code/__HELPERS/roundend.dm:188]:
```suggestion
	if(H.mind && (H.mind.special_role || H.mind.antag_datums?.len > 0))
```

File: code/__HELPERS/roundend.dm
Comment (spookydonut) [inline code/__HELPERS/roundend.dm:191]:
```suggestion
			for(var/i in A.objectives)
				var/datum/objective/O = i
```

**Review by spookydonut — CHANGES_REQUESTED** [review]
imagine actually compiling before you commit and push

**Comment by Qustinnus** [comment]
i did compile, it just failed and then i commited so i could go back to work kek

**Comment by Qustinnus** [comment]
else i wont have access to the changes on my other pc :^(

**Comment by anconfuzedrock** [comment]
ok but we don't _want_ people to fucking spawn as a blind warden before it was "Ok you can do this and not only will it  suck for you it will suck for the entire security team" now it's "Ok you can do this and it's the cool challenging way to play so you get gamer points while the rest of the security team suffers and gets nothing"

**Comment by Qustinnus** [comment]
i think you just have a problem with disabled people

**Comment by Moja10** [comment]
This reminds of blind ascension runs in nethack, and I think that's a great thing. But I understand where people are coming from worrying about bad quirk stacking now, as opposed to in the past, now that you've given it a gamer reward for doing so. Some people literally don't buy traitor items so they can have a tiny little sprite call them badass, in terms of small rewards encouraging players to act or choose a certain way.

**Comment by optimumtact** [comment]
I have asked the headadmins about if this should apply to team antag, security and command roles.

as such this PR is blocked until they reply

File: code/__HELPERS/roundend.dm
Comment (Rohesie) [inline code/__HELPERS/roundend.dm:188]:
```suggestion
	if(H.mind && (H.mind.special_role || length(H.mind.antag_datums) > 0))
```

File: code/__HELPERS/roundend.dm
Comment (Rohesie) [inline code/__HELPERS/roundend.dm:191]:
```suggestion
		for(var/a in H.mind.antag_datums)
			var/datum/antagonist/A = a
			for(var/i in A.objectives)
```

File: code/__HELPERS/roundend.dm
Comment (Rohesie) [inline code/__HELPERS/roundend.dm:468]: I'd check `GLOB.player_list` instead and check for being human and alive.
Way less mobs on that list, and it would avoid someone getting the message because someone dragged their body to shuttle.

File: code/__HELPERS/roundend.dm
Comment (Rohesie) [inline code/__HELPERS/roundend.dm:465]:
```suggestion
	. = list()
```

File: code/datums/traits/_quirk.dm
Comment (Rohesie) [inline code/datums/traits/_quirk.dm:13]:
```suggestion
	///Amount of points this trait is worth towards the hardcore character mode; minus points implies a positive quirk, positive means its hard. This is used to pick the quirks assigned to a hardcore character. 0 means its not available to hardcore draws.
	var/hardcore_value = 0
```

File: code/datums/traits/negative.dm
Comment (Rohesie) [inline code/datums/traits/negative.dm:341]:
```suggestion
	if(quirk_holder.client?.get_award_status(HARDCORE_RANDOM_SCORE) >= 5000) //More than 5k score? you unlock the gamer wheelchair.
```
Clients are incredibly volatile.

File: code/modules/client/preferences.dm
Comment (Rohesie) [inline code/modules/client/preferences.dm:1734]:
```suggestion
	if((randomise[RANDOM_SPECIES] || randomise[RANDOM_HARDCORE]) && !character_setup)

```

File: code/modules/client/preferences.dm
Comment (Rohesie) [inline code/modules/client/preferences.dm:1737]:
```suggestion
	if((randomise[RANDOM_BODY] || (randomise[RANDOM_BODY_ANTAG] && antagonist) || randomise[RANDOM_HARDCORE]) && !character_setup)

```

File: code/modules/client/preferences.dm
Comment (Rohesie) [inline code/modules/client/preferences.dm:1741]:
```suggestion
	if((randomise[RANDOM_NAME] || (randomise[RANDOM_NAME_ANTAG] && antagonist) || randomise[RANDOM_HARDCORE]) && !character_setup)

```

File: code/modules/client/preferences_savefile.dm
Comment (Rohesie) [inline code/modules/client/preferences_savefile.dm:142]: No need to add this. A failed access to an associative list should return a falsey value already.

File: code/modules/mob/dead/new_player/preferences_setup.dm
Comment (Rohesie) [inline code/modules/mob/dead/new_player/preferences_setup.dm:75]: Could you use more descriptive vars than `T`, `V`, `L`, `Q` and the likes?

File: code/modules/vehicles/wheelchair.dm
Comment (Rohesie) [inline code/modules/vehicles/wheelchair.dm:14]: I see this variable defined and re-defined, but not used. Is it WIP?
Document it, as well.

File: code/modules/client/preferences_savefile.dm
Comment (Qustinnus) [reply code/modules/client/preferences_savefile.dm:142]: alright, wasnt sure tbf working with prefs scares the fuck out of me and the same thing happens a few lines up so i figured better save than sorry

File: code/modules/vehicles/wheelchair.dm
Comment (Qustinnus) [reply code/modules/vehicles/wheelchair.dm:14]: Oh I thought I used it, I must've forgotten to, thanks.

File: code/modules/mob/dead/new_player/preferences_setup.dm
Comment (Qustinnus) [reply code/modules/mob/dead/new_player/preferences_setup.dm:75]: honestly I stole this from the original blacklist code which looks the exact same and I was scared of messing with it because it was hard to understand kek

**Comment by Qustinnus** [comment]
the headmins are not really replying and just keep saying theyll look into it but it's just a simple yes or no question

File: code/modules/mob/dead/new_player/preferences_setup.dm
Comment (Rohesie) [reply code/modules/mob/dead/new_player/preferences_setup.dm:75]: Just change the variable names for some that make sense, it won't affect logic.
It's really much less clear as it is, and understanding what this does is important.

**Comment by capn-monkeypaw** [comment]
> 
> 
> I have asked the headadmins about if this should apply to team antag, security and command roles.
> 
> as such this PR is blocked until they reply

Make exceptions for Team Antags and Command roles, please.

We're fine with Sec.

Thank you for waiting for our input.

File: code/modules/mob/dead/new_player/preferences_setup.dm
Comment (Qustinnus) [reply code/modules/mob/dead/new_player/preferences_setup.dm:75]: yea i was just explaining my reasoning kek, i changed it in the current commit

**Comment by Qustinnus** [comment]
I did all the requested changes, command and team antags can no longer get hardcore random. they can still easily get the exact same bad quirks if they pick them on their character like they could before, though

File: code/modules/vehicles/wheelchair.dm
Comment (Rohesie) [reply code/modules/vehicles/wheelchair.dm:14]: Could you document it?

File: code/modules/client/preferences_savefile.dm
Comment (Rohesie) [inline code/modules/client/preferences_savefile.dm:142]:
```suggestion
```
Just so there are no changes to the file.

File: code/modules/mob/dead/new_player/preferences_setup.dm
Comment (Rohesie) [inline code/modules/mob/dead/new_player/preferences_setup.dm:69]:
```suggestion
			var/datum/quirk/available_quirk = i
			if(available_hardcore_quirks[available_quirk] > quirk_budget)
				available_hardcore_quirks -= available_quirk
```

File: code/modules/mob/dead/new_player/preferences_setup.dm
Comment (Rohesie) [inline code/modules/mob/dead/new_player/preferences_setup.dm:73]:
```suggestion
		for(var/bl in SSquirks.quirk_blacklist) //Check if the quirk is blacklisted. quirk_blacklist is a list of lists.
```

File: code/modules/mob/dead/new_player/preferences_setup.dm
Comment (Rohesie) [inline code/modules/mob/dead/new_player/preferences_setup.dm:76]: `(picked_quirk in blacklist)` can be checked outside of the for loop, continuing early.

Comment (Qustinnus) [reply code/modules/mob/dead/new_player/preferences_setup.dm:76]: cool good idea I'll apply this to OG blacklist code too

File: code/controllers/subsystem/processing/quirks.dm
Comment (ExcessiveUseOfCobblestone) [inline code/controllers/subsystem/processing/quirks.dm:16]: doc?

File: code/controllers/subsystem/ticker.dm
Comment (Rohesie) [inline code/controllers/subsystem/ticker.dm:329]:
```suggestion
		if(H.mind?.special_role)
```

Comment (Rohesie) [reply code/controllers/subsystem/ticker.dm:329]: But even better, check the player list instead. This one can grow much bigger than that one.

Comment (Qustinnus) [reply code/controllers/subsystem/ticker.dm:329]: bruh :(

File: code/__HELPERS/roundend.dm
Comment (Rohesie) [inline code/__HELPERS/roundend.dm:477]:
```suggestion
		var/mob/living/carbon/human/human_player = i
		if(!human_player.hardcore_survival_score || !human_player.onCentCom() || human_player.stat == DEAD) ///gotta escape nerd
			continue
		if(!human_player.mind)
			continue
		. += "<li>[printplayer(human_player.mind)] with a hardcore random score of [round(human_player.hardcore_survival_score)]</li>"
```

File: code/__HELPERS/roundend.dm
Comment (Rohesie) [inline code/__HELPERS/roundend.dm:200]:
```suggestion
/datum/controller/subsystem/ticker/proc/HandleRandomHardcoreScore(client/player_client)
	if(!ishuman(player_client.mob))
		return FALSE
	var/mob/living/carbon/human/human_mob = player_client.mob
	if(!human_mob.hardcore_survival_score) ///no score no glory
		return FALSE

	if(human_mob.mind && (human_mob.mind.special_role || length(human_mob.mind.antag_datums) > 0))
		var/didthegamerwin = TRUE
		for(var/a in human_mob.mind.antag_datums)
			var/datum/antagonist/antag_datum = a
			for(var/i in antag_datum.objectives)
				var/datum/objective/objective_datum = i
				if(!objective_datum.check_completion())
					didthegamerwin = FALSE
		if(!didthegamerwin)
			return FALSE
		player_client.give_award(/datum/award/score/hardcore_random, human_mob, round(human_mob.hardcore_survival_score))
	else if(human_mob.onCentCom())
		player_client.give_award(/datum/award/score/hardcore_random, human_mob, round(human_mob.hardcore_survival_score))
```
Readable variables please!

Comment (Qustinnus) [reply code/__HELPERS/roundend.dm:200]: make this a contribution guideline please :( H and C for human and carbon have been allowed for years

File: code/controllers/subsystem/ticker.dm
Comment (Rohesie) [reply code/controllers/subsystem/ticker.dm:329]: ninjanomnom would 🗞️ you

Comment (Qustinnus) [reply code/controllers/subsystem/ticker.dm:329]: he's been approving my prs for years!! 😏

Comment (Rohesie) [reply code/controllers/subsystem/ticker.dm:329]: 😡

**Pull request merged**
